identity_credential/src/validator/domain_linkage_validator.rs

```diff
@@ -0,0 +1,530 @@
+      .linked_dids()
+      .iter()
+      .filter(|credential| credential.issuer.url().as_str() == issuer.did_str());
+
+    match matched_credentials.next() {
+      None => Err(DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::InvalidStructure,
+        source: None,
+      }),
+      Some(credential) => {
+        if matched_credentials.next().is_some() {
+          Err(DomainLinkageValidationError {
+            cause: DomainLinkageValidationErrorCause::InvalidStructure,
+            source: None,
+          })
+        } else {
+          Self::validate_credential(issuer, credential, domain.clone(), validation_options)
+        }
+      }
+    }
+  }
+
+  /// Validates a [Domain Linkage Credential](https://identity.foundation/.well-known/resources/did-configuration/#domain-linkage-credential).
+  ///
+  /// *`issuer`: issuer of the credential.
+  /// *`credential`: domain linkage Credential to be verified.
+  /// *`domain`: the domain hosting the credential.
+  pub fn validate_credential<T: Serialize, DOC: ValidatorDocument>(
+    issuer: &DOC,
+    credential: &Credential<T>,
+    domain: Url,
+    validation_options: &CredentialValidationOptions,
+  ) -> DomainLinkageValidationResult {
+    credential
+      .check_structure()
+      .map_err(|err| DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::CredentialValidationError,
+        source: Some(Box::new(err)),
+      })?;
+
+    let issuer_did: CoreDID =
+      CoreDID::parse(credential.issuer.url().to_string()).map_err(|err| DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::InvalidIssuer,
+        source: Some(Box::new(err)),
+      })?;
+
+    CredentialValidator::validate(credential, issuer, validation_options, FailFast::AllErrors).map_err(|err| {
+      DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::CredentialValidationError,
+        source: Some(Box::new(err)),
+      }
+    })?;
+
+    if credential.id.is_some() {
+      return Err(DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::ImpermissibleIdProperty,
+        source: None,
+      });
+    }
+
+    match &credential.types {
+      OneOrMany::Many(types) => {
+        if types.len() != 2 {
+          Err(DomainLinkageValidationError {
+            cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+            source: None,
+          })?;
+        }
+        let type_1 = types.get(0).ok_or(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+        let type_2: &String = types.get(1).ok_or(DomainLinkageValidationError {
+          cause: DomainLinkageValidationErrorCause::InvalidTypeProperty,
+          source: None,
+        })?;
+        let expected_types: HashSet<&str> = HashSet::from([
+          Credential::<T>::base_type(),
```

Comment:
Checking whether this is in the credential is part of `CredentialValidator::check_structure` which is part of `CredentialValidator::validate` already.